planner/core/logical_plan_builder.go

```diff
@@ -4613,38 +4641,60 @@ func (b *PlanBuilder) checkOriginWindowSpecs(funcs []*ast.WindowFuncExpr, orderB
 		if f.Spec.Name.L != "" {
 			spec = b.windowSpecs[f.Spec.Name.L]
 		}
-		if spec.Frame == nil {
-			continue
-		}
-		if spec.Frame.Type == ast.Groups {
-			return ErrNotSupportedYet.GenWithStackByArgs("GROUPS")
-		}
-		start, end := spec.Frame.Extent.Start, spec.Frame.Extent.End
-		if start.Type == ast.Following && start.UnBounded {
-			return ErrWindowFrameStartIllegal.GenWithStackByArgs(getWindowName(spec.Name.O))
-		}
-		if end.Type == ast.Preceding && end.UnBounded {
-			return ErrWindowFrameEndIllegal.GenWithStackByArgs(getWindowName(spec.Name.O))
-		}
-		if start.Type == ast.Following && (end.Type == ast.Preceding || end.Type == ast.CurrentRow) {
-			return ErrWindowFrameIllegal.GenWithStackByArgs(getWindowName(spec.Name.O))
-		}
-		if (start.Type == ast.Following || start.Type == ast.CurrentRow) && end.Type == ast.Preceding {
-			return ErrWindowFrameIllegal.GenWithStackByArgs(getWindowName(spec.Name.O))
+		if err := b.checkOriginWindowSpec(spec, orderByItems); err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		err := b.checkOriginWindowFrameBound(&start, spec, orderByItems)
+// checkWindowSpecs checks the validity for named window specifications.
+func (b *PlanBuilder) checkWindowSpecs(ctx context.Context, p LogicalPlan, aggMap map[*ast.AggregateFuncExpr]int) error {
+	for _, spec := range b.windowSpecs {
+		_, _, orderBy, _, err := b.buildProjectionForWindow(ctx, p, spec, nil, aggMap)
```

Comment:
Is there any method to avoid redundantly rebuild projection?

Reply:
Let me try another approach.